engine/src/main/java/io/zeebe/engine/processor/workflow/deployment/model/transformer/VariableMappingTransformer.java

```diff
@@ -0,0 +1,243 @@
+    final var context = new MappingContext();
+
+    for (final Mapping mapping : mappings) {
+      final var sourceExpression = mapping.source;
+      final var targetPathExpression = mapping.target;
+
+      final var targetPathParts = splitPathExpression(targetPathExpression);
+      createContextEntry(targetPathParts, sourceExpression, context);
+    }
+    return context;
+  }
+
+  private List<String> splitPathExpression(final String path) {
+    final var parts = path.split("\\.");
+    return new ArrayList<>(Arrays.asList(parts));
+  }
+
+  private void createContextEntry(
+      final List<String> targetPathParts,
+      final Expression sourceExpression,
+      final MappingContext context) {
+    final String target = targetPathParts.remove(0);
+
+    if (targetPathParts.isEmpty()) {
+      context.addEntry(target, sourceExpression);
+
+    } else {
+      final var nestedContext = context.getOrAddContext(target);
+      createContextEntry(targetPathParts, sourceExpression, nestedContext);
+    }
+  }
+
+  private String asFeelContextExpression(
+      final MappingContext context,
+      final BiFunction<String, List<String>, Object> contextValueVisitor) {
+    return context.visit(feelContextBuilder(contextValueVisitor));
+  }
+
+  private MappingContextVisitor<String> feelContextBuilder(
+      final BiFunction<String, List<String>, Object> contextValueVisitor) {
+    return new MappingContextVisitor<>() {
+      @Override
+      public String onEntry(final String targetKey, final Expression sourceExpression) {
+        return targetKey + ":" + sourceExpression.getExpression();
+      }
+
+      @Override
+      public String onContext(final List<String> entries) {
+        return "{" + String.join(",", entries) + "}";
+      }
+
+      @Override
+      public String onContextEntry(
+          final String targetKey, final String contextValue, final List<String> contextPath) {
+        return targetKey + ":" + contextValueVisitor.apply(contextValue, contextPath);
+      }
+    };
+  }
+
+  private String appendContextExpression(
+      final String nestedContext, final List<String> contextPath) {
+    // for a nested target mapping 'x -> a.b', append the nested property 'b' to
+    // the existing context variable 'a' (instead of overriding 'a')
+    // example: x = 1 and a = {'c':2} results in a = {'b':1, 'c':2}
+    final var existingContext = String.join(".", contextPath);
+    return "appendTo(" + existingContext + "," + nestedContext + ")";
+  }
+
+  private Expression parseExpression(
+      final String contextExpression, final ExpressionLanguage expressionLanguage) {
+    final var expression =
+        expressionLanguage.parseExpression(EXPRESSION_MARKER + contextExpression);
+
+    if (!expression.isValid()) {
+      throw new IllegalStateException(
+          String.format(
+              "Failed to build variable mapping expression: %s", expression.getFailureMessage()));
+    }
```

Comment:
No, I don't think so. There is no explicit failure handling for this case. I hope that this exception is never thrown - or only in a unit/integration test after a refactoring. 

The transformation of the variable mappings is done after the validation of the individual input/output mappings. To be safe, I could call the same logic in the validator and verify that the expression is valid.

Reply:
We don't expect it to happen due to invalid user input, right?
It would be safe as it is now. If transformation fails due to this exception, StreamProcessor detects the error and writes an error event. We don't crash or block other workflows. So it should be fine.

Reply:
Correct. The exception should only happen if there is a bug in the implementation. Invalid user input should be found during validation.